File: app/src/main/java/com/vamora/vano/data/VanoViewModel.kt
package com.vamora.vano.data

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class VanoViewModel(app: Application) : AndroidViewModel(app) {
    private val context = app.applicationContext
    val preferences = PreferencesManager(context)
    val chatRepository = ChatRepository(context)
    val modelManager = ModelManager(context)
    private val llama = LlamaInferenceManager()

    private val _selectedModel = MutableStateFlow<String?>(null)
    val selectedModel: StateFlow<String?> = _selectedModel

    private val _chatsLoaded = MutableStateFlow(false)
    val chatsLoaded: StateFlow<Boolean> = _chatsLoaded

    private val _isGenerating = MutableStateFlow(false)
    val isGenerating: StateFlow<Boolean> = _isGenerating

    private val _inferenceError = MutableStateFlow<String?>(null)
    val inferenceError: StateFlow<String?> = _inferenceError

    init {
        viewModelScope.launch {
            preferences.selectedModelFile.collectLatest { _selectedModel.value = it }
        }
        viewModelScope.launch {
            chatRepository.load()
            modelManager.refresh()
            // auto-select first available if none selected
            if (_selectedModel.value == null) {
                val available = modelManager.available.value
                if (available.isNotEmpty()) {
                    val first = available.first()
                    preferences.setSelectedModel(first.fileName)
                }
            }
            _chatsLoaded.value = true
        }
        viewModelScope.launch {
            modelManager.available.collectLatest { avail ->
                if (_selectedModel.value == null && avail.isNotEmpty()) {
                    preferences.setSelectedModel(avail.first().fileName)
                }
                if (_selectedModel.value != null && avail.none { it.fileName == _selectedModel.value }) {
                    // selected model removed, pick another if exists else null
                    val newPick = avail.firstOrNull()?.fileName
                    preferences.setSelectedModel(newPick)
                }
            }
        }
    }

    fun setSelectedModel(id: ModelId) {
        viewModelScope.launch {
            // unload previous to free RAM before switching
            if (_selectedModel.value != id.fileName) llama.unload()
            preferences.setSelectedModel(id.fileName)
        }
    }

    fun download(id: ModelId) {
        viewModelScope.launch { modelManager.download(id) }
    }

    fun cancelDownload(id: ModelId) = modelManager.cancel(id)

    fun deleteModel(id: ModelId) {
        viewModelScope.launch { modelManager.delete(id) }
    }

    fun deleteAllData(onDone: () -> Unit = {}) {
        viewModelScope.launch {
            llama.unload()
            modelManager.deleteAllAndClearData(chatRepository, preferences)
            onDone()
        }
    }

    fun createChat(onCreated: (String) -> Unit) {
        viewModelScope.launch {
            val chat = chatRepository.createChat()
            onCreated(chat.id)
        }
    }

    fun deleteChat(id: String) {
        viewModelScope.launch { chatRepository.deleteChat(id) }
    }

    fun deleteMessage(chatId: String, messageId: String) {
        viewModelScope.launch { chatRepository.deleteMessage(chatId, messageId) }
    }

    fun sendMessage(chatId: String, userText: String, attachmentName: String? = null, attachmentText: String? = null) {
        viewModelScope.launch {
            val displayText = if (userText.isBlank() && !attachmentText.isNullOrBlank()) "Attached ${attachmentName ?: "file"} for analysis" else userText
            val userMsg = ChatMessage(
                id = java.util.UUID.randomUUID().toString(),
                role = "user",
                content = userText.ifBlank { if (!attachmentText.isNullOrBlank()) "[File: $attachmentName]" else "" },
                timestamp = System.currentTimeMillis(),
                attachmentName = attachmentName,
                attachmentText = attachmentText?.take(15000)
            )
            chatRepository.addMessage(chatId, userMsg, newTitle = displayText.take(40))
            val selectedFile = _selectedModel.value
            if (selectedFile == null) {
                val err = ChatMessage(java.util.UUID.randomUUID().toString(), "assistant", "No model selected. Go to Settings and select a downloaded model.", System.currentTimeMillis())
                chatRepository.addMessage(chatId, err)
                return@launch
            }
            val modelId = ModelId.fromFileName(selectedFile)
            if (modelId == null || !modelManager.isDownloaded(modelId)) {
                val err = ChatMessage(java.util.UUID.randomUUID().toString(), "assistant", "Selected model not found. Please re-download it.", System.currentTimeMillis())
                chatRepository.addMessage(chatId, err)
                return@launch
            }
            _isGenerating.value = true
            _inferenceError.value = null
            // Create placeholder assistant message for streaming
            val assistantId = java.util.UUID.randomUUID().toString()
            val placeholder = ChatMessage(assistantId, "assistant", "▌", System.currentTimeMillis())
            chatRepository.addMessage(chatId, placeholder)

            try {
                val modelFile = modelManager.getModelFile(modelId)
                val loaded = llama.ensureLoaded(modelFile)
                if (!loaded) throw IllegalStateException("Failed to load GGUF (OOM? unsupported device). Try Vano mini.")

                // Build history before current user turn (exclude placeholder + current user msg)
                val chat = chatRepository.getChat(chatId)
                val history = chat?.messages?.filterNot { it.id == assistantId || it.id == userMsg.id } ?: emptyList()
                // For file analysis, append file block to prompt
                val promptWithFile = if (!attachmentText.isNullOrBlank()) {
                    "${userText.ifBlank { "Analyze this file" }}\n\n[Attached file: $attachmentName]\n${attachmentText.take(12000)}\n[End of file]"
                } else userText

                val modelTag = when (modelId) {
                    ModelId.VANO -> "vano Q4_0"
                    ModelId.VANO_MINI -> "vano Q3_K_M"
                }
                // Try real inference; on any native failure fallback to dummy so chat still works
                try {
                    val result = llama.generate(history, promptWithFile)
                    chatRepository.replaceMessageFull(
                        chatId, assistantId, result.text,
                        modelName = modelTag,
                        tokens = result.tokens,
                        genTimeMs = result.genTimeMs,
                        tokensPerSec = result.tokensPerSec
                    )
                } catch (e: Exception) {
                    _inferenceError.value = e.message
                    val label = modelId.displayName
                    val fallback = fallbackResponse(userText, label, e.message)
                    chatRepository.replaceMessageFull(chatId, assistantId, fallback, modelName = modelTag)
                }

            } catch (e: Exception) {
                chatRepository.replaceMessage(chatId, assistantId, "Error: ${e.message}")
                _inferenceError.value = e.message
            } finally {
                _isGenerating.value = false
            }
        }
    }

    private fun fallbackResponse(prompt: String, model: String, error: String?): String {
        return """$model (fallback) — you said: "$prompt"

On-device inference failed${if (error != null) ": $error" else ""}.

The model is downloaded but couldn't run (likely low RAM for the 1.55 GB Q4_0). Try "Vano mini" (1.38 GB) or close other apps and retry. Once loaded, answers are fully offline and private."""
    }

    override fun onCleared() {
        super.onCleared()
        viewModelScope.launch { llama.unload() }
    }
}

File: VanoIOS/ViewModels/VanoViewModel.swift
import Foundation
import Combine

@MainActor
class VanoViewModel: ObservableObject {
    @Published var selectedModel: String? = UserDefaults.standard.string(forKey: "selectedModel")
    @Published var isGenerating = false
    let modelManager = ModelManager()
    let chatStore = ChatStore()
    private let inference = InferenceManager()

    init() {
        chatStore.load()
        modelManager.refresh()
        if selectedModel == nil, let first = modelManager.available.first {
            selectedModel = first.fileName
            UserDefaults.standard.set(first.fileName, forKey: "selectedModel")
        }
        // Observe available to auto-select
    }

    func setSelectedModel(_ id: ModelId) {
        Task { await inference.unload() }
        selectedModel = id.fileName
        UserDefaults.standard.set(id.fileName, forKey: "selectedModel")
    }
    func createChat() -> Chat { chatStore.createChat() }
    func deleteChat(id: String) { chatStore.deleteChat(id: id) }
    func deleteMessage(chatId: String, messageId: String) { chatStore.deleteMessage(chatId: chatId, messageId: messageId) }
    func deleteAll() {
        Task { await inference.unload() }
        modelManager.deleteAll(chatStore: chatStore)
        selectedModel = nil
    }

    func sendMessage(chatId: String, text: String, fileName: String? = nil, fileText: String? = nil) {
        let displayTitle = text.isEmpty && fileText != nil ? "Attached \(fileName ?? "file")" : text
        let userMsg = ChatMessage(id: UUID().uuidString, role: "user", content: text.isEmpty && fileText != nil ? "[File: \(fileName ?? "file")]" : text, timestamp: Date(), attachmentName: fileName, attachmentText: fileText?.prefix(15000).description)
        chatStore.addMessage(chatId: chatId, message: userMsg, newTitle: String(displayTitle.prefix(40)))
        guard let sel = selectedModel, let modelId = ModelId(rawValue: sel), modelManager.isDownloaded(modelId) else {
            let err = ChatMessage(id: UUID().uuidString, role: "assistant", content: "No model selected. Go to Settings and select a downloaded model.", timestamp: Date())
            chatStore.addMessage(chatId: chatId, message: err)
            return
        }
        isGenerating = true
        let placeholderId = UUID().uuidString
        let placeholder = ChatMessage(id: placeholderId, role: "assistant", content: "▌", timestamp: Date())
        chatStore.addMessage(chatId: chatId, message: placeholder)
        Task {
            let fileURL = modelManager.fileURL(for: modelId)
            let loaded = await inference.ensureLoaded(fileURL: fileURL)
            guard loaded else {
                chatStore.updateMessage(chatId: chatId, messageId: placeholderId, content: "Error: Failed to load GGUF. Try Vano mini.")
                isGenerating = false
                return
            }
            let history = chatStore.chat(for: chatId)?.messages.filter { $0.id != placeholderId && $0.id != userMsg.id } ?? []
            let promptWithFile: String
            if let ft = fileText, !ft.isEmpty {
                promptWithFile = "\(text.isEmpty ? "Analyze this file" : text)\n\n[Attached file: \(fileName ?? "file")]\n\(ft.prefix(12000))\n[End of file]"
            } else { promptWithFile = text }
            do {
                let res = try await inference.generate(history: history, prompt: promptWithFile)
                chatStore.updateMessage(chatId: chatId, messageId: placeholderId, content: res.text, modelName: modelId.tag, tokens: res.tokens, genTimeMs: res.genTimeMs, tps: res.tokensPerSec)
            } catch {
                let fallback = "\(modelId.displayName) (fallback) — you said: \"\(text)\"\n\nOn-device inference failed: \(error.localizedDescription)"
                chatStore.updateMessage(chatId: chatId, messageId: placeholderId, content: fallback, modelName: modelId.tag)
            }
            isGenerating = false
        }
    }
}
